Add client option to disable following redirects

alprusty opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

During 412 release as part of OkHttp version upgrade, explicit client retry logic was removed in the event of a 307/308 redirect. This will cause 401: 'Unauthorized' error when trino client is talking to a Trino Cluster that is deployed behind a load balancer(For example AWS ALB) with a redirect mode(e.g. 307/308) routing policy.

Reference: https://github.com/trinodb/trino/commit/4be14018af945684bb28a53e4dcedfe744529bb0

Okhttp by default does follow redirects. However it drops Authorization headers if the redirect is across hosts. 

Reference: https://github.com/square/okhttp/blob/okhttp_4.10.x/okhttp/src/main/kotlin/okhttp3/internal/http/RetryAndFollowUpInterceptor.kt#L323-L328

Solution:
This PR introduces a flag `--disable-follow-redirects` to give an option to clients to follow a redirect or not. By default the flag is set to `false` which means redirects are going to be followed. 



<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

Fixes https://github.com/trinodb/trino/issues/21026

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

oneonestar commented:
We shouldn't make this change until we've a clear consensus on whether to include the Authorization header on redirects.
https://github.com/trinodb/trino/issues/21026#issuecomment-1991556960

mosabua commented:
Can you submit a CLA @alprusty ? @martint can process it soon after and trigger check here .. but we can work on reviews and more in parallel.

mosabua commented:
This was discussed with @electrum @dain @martint and others in the Trino Contributor Call from 21 March 2024.

alprusty commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -273,6 +273,9 @@ public class ClientOptions
     @Option(names = "--disable-auto-suggestion", description = "Disable auto suggestion")
     public boolean disableAutoSuggestion;
 
```
Introduced a new client option to allow/disallow redirects.

alprusty commented on client/trino-client/src/main/java/io/trino/client/OkHttpUtil.java:
```diff
@@ -88,14 +91,30 @@ public static Interceptor basicAuth(String user, String password)
+                    .build();
+
+            Response response = chain.proceed(request);
+
```
If `followRedirects` is enabled, then follow the redirect and reattach the `Authorization` header again while making the call.

oneonestar replied:
In Trino protocol, we sent the Authorization header with the request together. After received an redirect response, we should sent Authorization header in the redirect retry.
The `if redirect retry received 401 then retry with Authorization header` logic doesn't follow the current Trino protocol.
You might need to use [Network Interceptor](https://square.github.io/okhttp/features/interceptors/#network-interceptors).

alprusty replied:
@oneonestar 
One of the documentation https://square.github.io/okhttp/features/interceptors/#choosing-between-application-and-network-interceptors says Network interceptors are not invoked for cached response.

```
Not invoked for cached responses that short-circuit the network.
```

Wouldn't this cause issues to redirects if anything serverd via `v1/*` has a `Cache-control` header with `max-age` field ?

oneonestar replied:
https://square.github.io/okhttp/features/caching/
> OkHttp implements an optional, off by default,

I don't see Trino client or server is using the cache by okhttp.

mosabua commented:
CLA is signed .. I am working with @alprusty to get configuration fixed

mosabua commented:
@cla-bot check

oneonestar commented on docs/src/main/sphinx/client/cli.md:
```diff
@@ -149,6 +149,8 @@ mode:
 * - `--disable-auto-suggestion`
   - Disables autocomplete suggestions.
+* - `--follow-redirects`
+  - If set to true, then follows client redirects. Default is false.
```
According to the discussion in #21026, we agree to set this to true by default.

alprusty replied:
Sure. Updating the PR to make it `true` by default.

oneonestar commented:
@alprusty We also need a regression test on the client actually sent the Authorization header after a redirect.

alprusty commented:
@oneonestar 
I took care of all the review comments.
- Made `followRedirects` default to `true` instead of `false`
- Added Network interceptor to OkHttp builder
- Added a test case to verify Authorization header is being pushed on a redirect https://github.com/trinodb/trino/blob/4dffa6b6d645e26581b5992a8a3caff198446883/client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java

Please review again 🙏

oneonestar commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -273,6 +273,9 @@ public class ClientOptions
     @Option(names = "--disable-auto-suggestion", description = "Disable auto suggestion")
     public boolean disableAutoSuggestion;
 
+    @Option(names = "--follow-redirects", description = "Follow client redirects from server")
```
Since redirect is default true, we might want a flag to disable the it.
Following the current naming convention, I suggest `--disable-follow-redirect`.

alprusty replied:
Ack.

oneonestar commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -273,6 +273,9 @@ public class ClientOptions
     public boolean disableAutoSuggestion;
 
+    @Option(names = "--follow-redirects", description = "Follow client redirects from server")
+    public Optional<Boolean> followRedirects;
```
Can this be `boolean`?

alprusty replied:
Yeah, with the positive flag naming convention `null/empty` or `true` meant flag is set,  and explicit `false` meant flag is disabled. Hence went with `Boolean` instead of `boolean`

With a negative flag name e.g. `--disable-follow-redirect`, we can use `boolean` instead of `Boolean`

sajjoseph commented:
It will be great if this PR can be split into two.

1. Control `followRedirects` in okhttp
2. Retain authorization header during redirects

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    @Test
+    public void testDisableClientRedirectsDefault()
```
This test still pass when I modify `addNetworkInterceptor()` to `addInterceptor()`.
I think the test has to be redirect to another host:port.
Maybe one MockWebServer for proxy and one for Trino Coordinator.

alprusty replied:
Added two mock servers, one for proxy one for coordinator

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+        }
+
+        //verify the redirect request has the same Authorization header
+        RecordedRequest redirectedRequest = server.takeRequest();
+        assertThat(redirectedRequest.getHeader(AUTHORIZATION)).isEqualTo("Bearer " + DUMMY_ACCESS_TOKEN);
```
This tests the first request (to proxy), which always contain the AUTHORIZATION header.
We want the second request (to Trino) to contain the AUTHORIZATION header.

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+        if (disableFollowRedirects.isPresent()) {
+            properties.setProperty(PropertyName.DISABLE_FOLLOW_REDIRECTS.toString(), Boolean.toString(disableFollowRedirects.get()));
+        }
+        properties.setProperty(PropertyName.ACCESS_TOKEN.toString(), DUMMY_ACCESS_TOKEN);
```
Test the following:
password (`Authorization: Basic <credentials>`)
accessToken, externalAuthentication (`Authorization: Bearer <token>`)

Network interceptor for password is missing in the current implementation.

alprusty replied:
Added test cases for `basicAuth`, `accessToken` and `externalAuth`

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+    public void testClientRedirectEnabled()
+            throws Exception
+    {
+        server.enqueue(new MockResponse()
+                .setResponseCode(307)
+                .addHeader(LOCATION, server.url("redirect/v1/statement"))
+                .addHeader(AUTHORIZATION, DUMMY_ACCESS_TOKEN));
+        server.enqueue(new MockResponse()
+                .addHeader(CONTENT_TYPE, "application/json")
+                .setBody(createResults(server)));
+
+        QueryRunner queryRunner = createQueryRunner(createTrinoUri(server, Optional.of(false)), createClientSession(server));
+
+        try (Query query = queryRunner.startQuery("SELECT 1")) {
+            query.renderOutput(getTerminal(), nullPrintStream(), nullPrintStream(), CSV, Optional.of(""), false);
+        }
+
+        //verify the redirect request has the same Authorization header
+        RecordedRequest redirectedRequest = server.takeRequest();
+        assertThat(redirectedRequest.getHeader(AUTHORIZATION)).isEqualTo("Bearer " + DUMMY_ACCESS_TOKEN);
```
The test in this function should be exactly the same as `testDisableClientRedirectsDefault()` except QueryRunner.
Extract function to prevent code duplication.

oneonestar commented on client/trino-client/src/main/java/io/trino/client/JsonResponse.java:
```diff
@@ -140,6 +141,11 @@ public static <T> JsonResponse<T> execute(JsonCodec<T> codec, Call.Factory clien
                 return new JsonResponse<>(response.code(), response.headers(), body, value, exception);
             }
+
+            if (response.isRedirect() && !((OkHttpClient) client).followRedirects()) {
+                throw new ClientException("Client redirection was detected, but 'disableFollowRedirects' flag is set to true");
+            }
```
I don't feed good about this downcasting. Also the check should be handled by upper layer, not inside JsonResponse.

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+        QueryRunner queryRunner = createQueryRunner(createTrinoUri(server, Optional.empty()), createClientSession(server));
+
+        try (Query query = queryRunner.startQuery("SELECT 1")) {
+            query.renderOutput(getTerminal(), nullPrintStream(), nullPrintStream(), CSV, Optional.of(""), false);
```
Can we assert the query result instead of trying to render the result? Same for others.

alprusty replied:
Could you please share an example unit test where this is happening, couldn't find one.

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,191 @@
+    {
+        return new PrintStream(nullOutputStream());
+    }
+}
```
I would like another test for multiple redirection.

alprusty replied:
`testClientRedirectsMultipleHop`

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,316 @@
+        }
+    }
+
+    static TrinoUri createTrinoUriAccessToken(MockWebServer server, Optional<Boolean> disableFollowRedirects)
```
`createTrinoUriAccessToken`, `createTrinoUriBasicAuth` and `createTrinoUriExternalAuth` are mostly identical.
We should extract a helper function.

oneonestar commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,316 @@
+                HttpLoggingInterceptor.Level.NONE);
+    }
+
+    static PrintStream nullPrintStream()
```
Add `private` to these functions.

oneonestar commented on docs/src/main/sphinx/client/cli.md:
```diff
@@ -149,6 +149,8 @@ mode:
 * - `--disable-auto-suggestion`
   - Disables autocomplete suggestions.
+* - `--disable-follow-redirects`
+  - If set to `true`, then disables client redirects. Default is `false`.
```
@mosabua Any comments for the docs?

mosabua replied:
```suggestion
  - Set to `true` to turn off client redirects. Defaults to `false`.
```

oneonestar reviewed (commented):
LGTM except some minor comments.

@dain @electrum Please could you take a look.
I hope we can land this in the next release, since #21203 is only an incomplete fix.

alprusty commented on client/trino-cli/src/test/java/io/trino/cli/TestClientRedirect.java:
```diff
@@ -0,0 +1,316 @@
+
+    @Test
+    public void testDisableClientRedirectsDefaultWithAccessToken()
+            throws Exception
+    {
+        assertRedirectIsFollowedWithAuthorizationHeader(Optional.empty(), false);
+    }
+
+    @Test
+    public void testClientRedirectEnabledWithAccessToken()
+            throws Exception
+    {
+        assertRedirectIsFollowedWithAuthorizationHeader(Optional.of(false), false);
+    }
+
+    @Test
+    public void testClientRedirectDisabledWithAccessToken()
+    {
+        assertRedirectIsNotFollowed(false);
+    }
+
+    @Test
+    public void testDisableClientRedirectsDefaultWithBasicAuth()
+            throws Exception
+    {
+        assertRedirectIsFollowedWithAuthorizationHeader(Optional.empty(), true);
+    }
+
+    @Test
+    public void testClientRedirectEnabledWithBasicAuth()
+            throws Exception
+    {
+        assertRedirectIsFollowedWithAuthorizationHeader(Optional.of(false), true);
+    }
+
+    @Test
+    public void testClientRedirectDisabledWithBasicAuth()
+    {
+        assertRedirectIsNotFollowed(true);
+    }
```
there is a bit of setup() and teardown() flow going on to reset mock servers between methods. If we combine these methods, then have to add that logic inside. Hence skipping this.

alprusty committed:
add client option to disable following redirects

alprusty commented:
squashed all commits as one

mosabua commented on docs/src/main/sphinx/client/jdbc.md:
```diff
@@ -250,3 +250,6 @@ may not be specified using both methods.
     network overhead and uses smaller HTTP headers and requires Trino 431 or
     greater.
+* - `disableFollowRedirects`
+  - If set to `true`, then disables client redirects. Default is `false`.
```
```suggestion
  - Set to `true` to turn off client redirects. Defaults to `false`.
```

alprusty replied:
Updated.

mosabua commented:
We still want to continue with this. Can you rebase @alprusty

mhassanzahid commented:
@alprusty @mosabua is this still going forward or has this been added in some other release?
